src/lib/shell/commands/Ota.cpp

```diff
@@ -0,0 +1,268 @@
+{
+    // TODO: Implement a correct URI parser
+    constexpr char protocol[]    = "bdx://";
+    constexpr size_t protocolLen = sizeof(protocol) - 1;
+
+    if (imageURI.size() >= protocolLen && strncmp(protocol, imageURI.data(), imageURI.size()) == 0)
+    {
+        imageURI = imageURI.SubSpan(protocolLen, imageURI.size() - protocolLen);
+    }
+
+    for (size_t i = 0; i < imageURI.size(); ++i)
+    {
+        if (imageURI.data()[i] == '/')
+        {
+            return imageURI.SubSpan(i + 1, imageURI.size() - i - 1);
+        }
+    }
+
+    return imageURI;
+}
+
+void OnQueryImageResponse(void * context, const QueryImageResponse::DecodableType & response)
+{
+    ChipLogProgress(SoftwareUpdate, "Received QueryImage response: %" PRIu16, static_cast<uint16_t>(response.status));
+
+    CharSpan imageURI     = response.imageURI.ValueOr({});
+    CharSpan resourceName = ExtractResourceName(imageURI);
+    ChipLogProgress(SoftwareUpdate, "  Image URI: %.*s", static_cast<int>(imageURI.size()), imageURI.data());
+    ChipLogProgress(SoftwareUpdate, "  Resource: %.*s", static_cast<int>(resourceName.size()), resourceName.data());
+
+    OperationalDeviceProxy * deviceProxy = Server::GetInstance().GetOperationalDeviceProxy();
+    VerifyOrReturn(deviceProxy != nullptr);
+
+    Messaging::ExchangeManager * exchangeMgr = deviceProxy->GetExchangeManager();
+    Optional<SessionHandle> session          = deviceProxy->GetSecureSession();
+    Messaging::ExchangeContext * exchangeCtx = nullptr;
+
+    if (exchangeMgr != nullptr && session.HasValue())
+    {
+        exchangeCtx = exchangeMgr->NewContext(session.Value(), &sBdxDownloader);
+    }
+
+    if (exchangeCtx == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "Failed to allocate exchange");
+        return;
+    }
+
+    TransferSession::TransferInitData initOptions;
+    initOptions.TransferCtlFlags = TransferControlFlags::kReceiverDrive;
+    initOptions.MaxBlockSize     = 1024;
+    initOptions.FileDesLength    = resourceName.size();
+    initOptions.FileDesignator   = reinterpret_cast<const uint8_t *>(resourceName.data());
+
+    sBdxDownloader.SetInitialExchange(exchangeCtx);
+    CHIP_ERROR error = sBdxDownloader.InitiateTransfer(&DeviceLayer::SystemLayer(), TransferRole::kReceiver, initOptions,
+                                                       System::Clock::Seconds16(20));
+
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(SoftwareUpdate, "Failed to initiate BDX transfer: %" CHIP_ERROR_FORMAT, error.Format());
+    }
+}
+
+void OnQueryImageFailure(void * /* context */, EmberAfStatus status)
+{
+    ChipLogError(SoftwareUpdate, "QueryImage failed: %" PRIu16, static_cast<uint16_t>(status));
+}
+
+void OnQueryImageConnection(void * /* context */, DeviceProxy * deviceProxy)
+{
+    // Initialize cluster object
+    Controller::OtaSoftwareUpdateProviderCluster cluster;
+    CHIP_ERROR error = cluster.Associate(deviceProxy, kProviderEndpointId);
```

Comment:
The `kProviderEndpointId` is not going to be fixed on all providers. Add a TODO that the location of the endpoint needs to be found.

Reply:
Isn't OTA Provider cluster going to be a utility cluster, that is, part of the root node device type? If not, how will OTA Requestor find the endpoint ID? Using Descriptor cluster? `AnnounceOTAProvider` doesn't contain endpoint ID either.

Reply:
It was an omission that the DefaultOTAProviders list and AnnounceOTAProvider don't record the endpoint to use.

OTAProvider is NOT on the RootNode device type, and neither is OTARequestor, for that matter. Both of those are Device Types that could be on any endpoint

Reply:
Good to know. Added endpoint-id as another command argument.